### PR TITLE
DEV: convert post comments widgets to Glimmer components

### PR DESCRIPTION
**Pull request by renato**
This PR converts all post comments widgets (`post-voting-comments` and all widgets within) to Glimmer components, leaving only `post-voting-post` as a widget. I tried not to differ too much from the original structure to avoid regressions, especially on custom themes.

The main intent of this work is to allow the plugin to still work correctly with browser extensions that manipulate the DOM, like Grammarly, which currently breaks the event handling and makes it unusable.

It also adds a shim to `post-voting-button` to render the Glimmer component when used by the `post-voting-post` widget and to `post-voting-comments` to render the component when decorating `post-menu:after`.

### CODE REVIEW

**Review by xfalcox — APPROVED** [review]

File: assets/javascripts/discourse/components/post-voting-comment-composer.hbs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+          count=this.siteSettings.min_post_length
+        }}
+      </div>
+    {{else if this.isValueTooLong}}
+      <div class="post-voting-comment-composer-flash error">
+        {{i18n
+          "post_voting.post.post_voting_comment.composer.too_long"
+          count=this.siteSettings.qa_comment_max_raw_length
+        }}
+      </div>
```
Comment (megothss) [inline assets/javascripts/discourse/components/post-voting-comment-composer.hbs:23]: Wouldn't it be better to set which error should be displayed in a JS getter and avoid the `else if` here in the template?

File: assets/javascripts/discourse/components/post-voting-comment-composer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  @action
+  onInput(event) {
+    this.value = event.target.value;
+    this.args.onInput?.(event.target.value);
+  }
+
+  get isValueTooShort() {
+    return (
+      this.value.length > 0 &&
+      this.value.length < this.siteSettings.min_post_length
```
Comment (megothss) [inline assets/javascripts/discourse/components/post-voting-comment-composer.js:20]: I'm not convinced, if it is worth bothering with this edge case here, but what if `this.siteSettings.min_post_length` > `this.siteSettings.qa_comment_max_raw_length`?

File: assets/javascripts/discourse/components/post-voting-comment-editor.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+export default class PostVotingCommentEditor extends Component {
+  @service siteSettings;
+
+  @tracked value = this.args.raw;
+  @tracked submitDisabled = true;
+
+  @action
+  updateValue(value) {
+    this.value = value;
+    this.submitDisabled =
```
Comment (megothss) [inline assets/javascripts/discourse/components/post-voting-comment-editor.js:17]: The same, what if `this.siteSettings.min_post_length` > `this.siteSettings.qa_comment_max_raw_length`?

File: assets/javascripts/discourse/components/post-voting-comments-menu.hbs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      @action={{if
+        this.currentUser
+        this.expandComposer
+        (route-action "showLogin")
+      }}
+      @label="post_voting.post.post_voting_comment.add"
+    />
+  {{/if}}
+
+  {{#if (gt @moreCommentCount 0)}}
```
Comment (megothss) [inline assets/javascripts/discourse/components/post-voting-comments-menu.hbs:21]: I would avoid comparing the values like this in the handlebars template and move it to a getter property in the JS file.

It makes the template harder to read.

File: assets/javascripts/discourse/components/post-voting-button.hbs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<button
```
Comment (megothss) [inline assets/javascripts/discourse/components/post-voting-button.hbs:1]: Any reason to not use `DButton` here?

**Review by megothss — APPROVED** [review]
Looks good! Just some minor nitpicks.

File: assets/javascripts/discourse/components/post-voting-comment-composer.js (hunk 2 above)
Comment (renato) [reply assets/javascripts/discourse/components/post-voting-comment-composer.js:20]: I don't think it's worth it, `qa_comment_max_raw_length` is a hidden setting with a default of 600 (not sure why it's hidden though) and the error messages are a good hint on how to "solve" it if it happens

File: assets/javascripts/discourse/components/post-voting-comment-editor.js (hunk 3 above)
Comment (renato) [reply assets/javascripts/discourse/components/post-voting-comment-editor.js:17]: I don't think it's worth it, `qa_comment_max_raw_length` is a hidden setting with a default of 600 (not sure why it's hidden though) and the error messages are a good hint on how to "solve" it if it happens

File: assets/javascripts/discourse/components/post-voting-button.hbs (hunk 5 above)
Comment (renato) [reply assets/javascripts/discourse/components/post-voting-button.hbs]: Done!

File: assets/javascripts/discourse/components/post-voting-comments-menu.hbs (hunk 4 above)
Comment (renato) [reply assets/javascripts/discourse/components/post-voting-comments-menu.hbs]: Agreed, done!

File: assets/javascripts/discourse/components/post-voting-comment-composer.hbs (hunk 1 above)
Comment (renato) [reply assets/javascripts/discourse/components/post-voting-comment-composer.hbs]: Yep, sounds good, changed to an `errorMessage` getter.

**Comment by renato** [comment]
Thank you both for the reviews!

**Comment by notz** [comment]
Should this plugin be compatible with 3.0.6 version?

Because after updating i get following javascript errors, which breaks discourse from loading? 

```
Uncaught (in promise) TypeError: (0 , e.registerWidgetShim) is not a function
    at s.callback (post-voting-button-shim.js:5:1)
    at s.exports (loader.js:106:1)
    at requireModule (loader.js:27:1)
    at auto-load-modules.js:16:1
    at Array.forEach (<anonymous>)
    at s (auto-load-modules.js:11:1)
    at Object.initialize (auto-load-modules.js:37:1)
    at o.initialize (app.js:41:1)
    at index.js:126:1
    at e.each (dag-map.js:192:1)
    at e.walk (dag-map.js:121:1)
    at e.each (dag-map.js:66:1)
    at e.topsort (dag-map.js:72:1)
    at e._runInitializer (index.js:138:1)
    at e.runInstanceInitializers (index.js:124:1)
    at e._bootSync (instance.js:101:1)
    at e.didBecomeReady (application.js:650:1)
    at p.invoke (queue.ts:201:14)
    at p.flush (queue.ts:98:13)
    at h.flush (deferred-action-queues.ts:75:19)
    at q._end (index.ts:616:32)
    at _boundAutorunEnd (index.ts:257:12)
```

```
define("discourse/plugins/discourse-post-voting/discourse/widgets/post-voting-button-shim", ["@ember/template-factory", "discourse/widgets/render-glimmer"], function (_templateFactory, _renderGlimmer) {
  "use strict";

  0; //eaimeta@70e063a35619d71f0,"ember-cli-htmlbars",0,"discourse/widgets/render-glimmer"eaimeta@70e063a35619d71f
  (0, _renderGlimmer.registerWidgetShim)("post-voting-button", "div.post-voting-button-shim", (0, _templateFactory.createTemplateFactory)(
  /*
    <PostVotingButton
      @direction={{@data.direction}}
      @loading={{@data.loading}}
      @voted={{@data.voted}}
      @removeVote={{@data.removeVote}}
      @vote={{@data.vote}}
    />
  */
  {
    "id": "A1R6Vd4j",
    "block": "[[[8,[39,0],null,[[\"@direction\",\"@loading\",\"@voted\",\"@removeVote\",\"@vote\"],[[30,1,[\"direction\"]],[30,1,[\"loading\"]],[30,1,[\"voted\"]],[30,1,[\"removeVote\"]],[30,1,[\"vote\"]]]],null]],[\"@data\"],false,[\"post-voting-button\"]]",
    "moduleName": "(unknown template module)",
    "isStrictMode": false
  }));
});
```

**Comment by xfalcox** [comment]
> Should this plugin be compatible with 3.0.6 version?

All our plugins track our default release channel, which is tests-passed.

If you are on stable you can pin to an older commit until next week when we bump stable.

**Comment by notz** [comment]
@xfalcox Thank's for explaination. Then it seems to be hard do not run in troubles with docker/kubernetes installations, because they follow the stable channel. Shouldn't be the ".discourse-compatibility" file be updated in this case?

**Pull request merged**